### PR TITLE
Manage and use own storage pool

### LINKED ISSUE #1457: Use own storage pool
Currently Installer assumes a `default` storage pool, which is not guaranteed and hence the reason for [documenting the case of it not existing](https://github.com/openshift/installer/blob/master/docs/dev/libvirt-howto.md#configure-default-libvirt-storage-pool). I think it'd be best if Installer used its own custom storage pool.

Before this can be fixed, these two will have to be resolved first:

- [x]  #308 through https://github.com/dmacvicar/terraform-provider-libvirt/issues/435
- [x]  https://github.com/openshift/cluster-api-provider-libvirt/pull/45

I'll be working on these and this issue itself so please consider this more like a TODO item for me. :)

# Platform (aws|libvirt|openstack): libvirt

### PR DESCRIPTION
**Pull request by zeenix**
Don't assume `default` storage pool and create a new custom one.

Fixes #1457.

### CODE REVIEW

**Comment by zeenix** [comment]
gah, that's not at all a useful log from `tf-fmt`: https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_installer/1956/pull-ci-openshift-installer-tf-fmt/5020 :(

**Comment by zeenix** [comment]
/assign @skitt

Could you do a review @skitt please?

File: data/data/libvirt/main.tf [hunk 1]
```diff
@@ -2,11 +2,18 @@ provider "libvirt" {
   uri = var.libvirt_uri
 }
 
+resource "libvirt_pool" "storage_pool" {
+  name = var.cluster_id
+  type = "dir"
+  path = "/var/lib/libvirt/images/${var.cluster_id}"
```
Comment (skitt) [inline data/data/libvirt/main.tf]: This ends up inside the default pool, if it’s defined following the default configuration; is that what we want? Shouldn’t we use another path, *e.g.* `/var/lib/libvirt/openshift/${var.cluster_id}`?

Comment (zeenix) [reply data/data/libvirt/main.tf]: good catch! yeah i think we should not be under the default pool's dir.

Comment (zeenix) [reply data/data/libvirt/main.tf]: Updated! I didn't test but that's what the CI is for. :)

Comment (skitt) [reply data/data/libvirt/main.tf]: Tested, works for me:

    DEBUG libvirt_volume.master[0]: Creation complete after 18s [id=/var/lib/libvirt/openshift-images/test1-dj2nv/test1-dj2nv-master-0]

This is with SELinux on too ;-).

**Comment by zeenix** [comment]
/test e2e-aws

Also as per bot's orders :)

/assign @abhinavdahiya

**Comment by abhinavdahiya** [comment]
74546e66251f702287fdcdff36153f05bc9d5bfb is
/approve

**Comment by zeenix** [comment]
seems it still needs `/lgtm`. @skitt could you give that?

**Comment by zeenix** [comment]
> [74546e6](https://github.com/openshift/installer/commit/74546e66251f702287fdcdff36153f05bc9d5bfb) is
> /approve

Thanks. Do you mean only that commit is approved?

**Comment by skitt** [comment]
> seems it still needs `/lgtm`. @skitt could you please give that?

I’m not sure mine counts, but here goes:

/lgtm

**Comment by zeenix** [comment]
@skitt thanks, obviously it does. :)

/test e2e-aws

**Comment by skitt** [comment]
> @skitt thanks, obviously it does. :)

Yup, I need to flex my muscles a bit more ;-).

**Comment by markmc** [comment]
/hold

Sorry I'm late to this, but since the `baremetal` platform also uses libvirt (for the bootstrap VM), we will probably want to follow whatever you do here

I completely agree with the premise of this - users should not have to manually create the "default" storage pool before running the installer

But consider this - it has always been intended that libvirt installation would result in the `/var/lib/libvirt/images` path being created and then the first user of that would create the 'default' storage pool pointing to that. Here's what virt-manager and virt-install do:

https://www.redhat.com/archives/libvir-list/2008-August/msg00179.html
https://github.com/virt-manager/virt-manager/blob/2b24a85884f8242ab2ade14e6403d447be07f0f1/virtinst/storage.py#L154-L163

(Whether this was a good idea or not is another matter - I'm just explaining the intent!)

My conclusion - openshift-install should follow this approach and just create the default storage pool if it doesn't exist.

Why? I think `/var/lib/libvirt/images` is a well known location (and e.g. people are used to it containing large VM images) so I think it would be better for us to continue using that rather than switch to `/var/lib//libvirt/openshift-images`

**Comment by zeenix** [comment]
@markmc I created https://github.com/openshift/installer/issues/1457 months ago and spent considerable amount of time implementing storage pools in terraform-provider-libvirt. In my discussion with Openshift devs, nobody objected to this before either so can you imagine my frustration on you blocking this at the last mile of a rather long journey?

People expecting images to be under a very specific location is not a great argument IMO and certainly not a good one to present this late. I already updated the code to not assume a specific path for images:

https://github.com/openshift/cluster-api-provider-libvirt/pull/144
https://github.com/openshift/installer/pull/1628

If your code assumes a path, please update it as that is not the correct thing to do. As mentioned before `default` storage pool is not guaranteed to exist to assuming this pool's existance is not correct from libvirt's POV.

As for the need for this, this is part of my effort to hopefully move us to the session libvirt connection by default or at least be able to use session libvirt if user asks for it. Currently Installer happily let's you provide a libvirt connection URL but it can't actually handle anything other than the system libvirt connection. One reason that is the case is assumption of default storage pool in a specific location.

**Comment by zeenix** [comment]
>  Here's what virt-manager and virt-install

Yes, that's where the assumption of `default` storage pool at a specific location comes from since most libvirt users use either or both of these tools but there is no reason to believe that is all the users. There is a reason why libvirt:

1. doesn't create the `default` pool for you.
2. has the storage pool creation/management API and not just volumes.

**Comment by zeenix** [comment]
> openshift-install should follow this approach and just create the default storage pool if it doesn't exist.

I actually thought about that but how do you implement this? Can you tell TF to not fail if the resource already exists and not manage it? I couldn't think of any easy solution there and since every libvirt client should use it's own storage pool, I didn't see investing time on solving this non-problem.

**Comment by markmc** [comment]
The hold was to ask you to give the point some consideration before merging, that's all

/hold cancel

**Comment by zeenix** [comment]
@markmc oh ok. I'm terribly sorry that I ended up sounding so rude. It wasn't intentional.

**Pull request merged**